opttoy2/opttoy.go

```diff
@@ -0,0 +1,1120 @@
+		}
+
+	case parser.UnqualifiedStar:
+		e.inputVars = parent.inputVars
+
+	case parser.UnresolvedName:
+		if len(b) != 2 {
+			fatalf("unsupported unqualified name: %s", b)
+		}
+		tableName := string(b[0].(parser.Name))
+		colName := string(b[1].(parser.Name))
+		if base, ok := state.tables[tableName]; !ok {
+			fatalf("unknown table %s", b)
+		} else if table, ok := state.catalog[tableName]; !ok {
+			fatalf("unknown table %s", b)
+		} else if colIndex, ok := table.columns[colName]; !ok {
+			fatalf("unknown column %s", b)
+		} else {
+			e.inputVars.set(base + bitmapIndex(colIndex))
+		}
+
+	case *parser.NumVal:
+
+	case parser.NaturalJoinCond:
+		e.resolveNaturalJoin(state)
+
+	case *parser.UsingJoinCond:
+		e.resolveUsingJoin(state, b.Cols)
+
+	case *parser.ExistsExpr:
+		// TODO(peter): unimplemented.
+
+	default:
+		unimplemented("%T", e.body)
+	}
+}
+
+func (e *expr) resolveNaturalJoin(state *queryState) {
+	common := make(map[string]struct{})
+	for i, input := range e.inputs() {
+		table := input.body.(*table)
+		if i == 0 {
+			for col := range table.columns {
+				common[col] = struct{}{}
+			}
+		} else {
+			for col := range common {
+				if _, ok := table.columns[col]; !ok {
+					delete(common, col)
+				}
+			}
+		}
+	}
+
+	names := make(parser.NameList, 0, len(common))
+	for col := range common {
+		names = append(names, parser.Name(col))
+	}
+	sort.Slice(names, func(i, j int) bool {
+		return names[i] < names[j]
+	})
+
+	e.resolveUsingJoin(state, names)
+}
+
+func (e *expr) resolveUsingJoin(state *queryState, names parser.NameList) {
+	// TODO(peter): check for validity of the names.
+
+	for _, name := range names {
+		// For every adjacent pair of tables, add an equality predicate.
+		inputs := e.inputs()
+		for i := 1; i < len(inputs); i++ {
+			left := inputs[i-1].body.(*table)
+			right := inputs[i].body.(*table)
+			e.addFilter(&expr{
+				op: eqOp,
+				children: []*expr{
+					{
+						op: variableOp,
+						body: parser.UnresolvedName{
+							parser.Name(left.name),
+							name,
+						},
+					},
+					{
+						op: variableOp,
+						body: parser.UnresolvedName{
+							parser.Name(right.name),
+							name,
+						},
+					},
+				},
+				inputCount: 2,
+			})
+		}
+	}
+
+	e.body = nil
+}
+
+func (e *expr) substitute(columns bitmap, replacement *expr) *expr {
+	if e.op == variableOp && e.outputVars == columns {
+		return replacement
+	}
+
+	result := *e
+	result.children = append([]*expr(nil), e.children...)
+
+	for i, input := range result.inputs() {
+		result.children[i] = input.substitute(columns, replacement)
+	}
+	result.updateProperties()
+	return &result
+}
+
+func (e *expr) isFilterCompatible(filter *expr) bool {
+	// NB: when pushing down a filter, the filter applies before the projection
+	// and thus needs to be compatible with the input variables, not the output
+	// variables.
+	return (filter.inputVars & e.inputVars) == filter.inputVars
+}
+
+func buildEquivalencyMap(filters []*expr) map[bitmap]*expr {
```

Comment:
This map could use some comments. We are mapping sets of variables to equivalent expressions?

It seems off that we are treating `(a, b) = (c, d)` differently than `a = b AND b = c` (the equivalency map is different between these two). If we have filters `a = e`, `b = f`, and `(c, d) = (a, b)` (with one input providing `a, b` and another providing `c, d,e,f`), we won't be able to infer `(c, d) = (e, f)`

It seems that the right thing to do here is to create equivalency groups between variables (e.g. UnionFind) and then trying to find a variable in each group that is in `filter.inputVars`

Reply:
Definitely any node corresponding to a relational operator should have a set of physical properties (which includes equivalency groups).

> if you find the manipulations in this PR more readable than those present in the existing Cockroach code base.

Absolutely. This seems much nicer and easy to understand. Though there may be downsides I'm not seeing right now.

One thing I'm not quite clear how it will work is when we have rendered expressions (e.g. an `a+b` somewhere in there). Would these show up as special variables in our structures?

Reply:
> Definitely any node corresponding to a relational operator should have a set of physical properties (which includes equivalency groups).

In my discussions with Alberto, he's indicated that we should separate the logical and physical properties. The physical properties only arise when we get to physical nodes such as index scans. Though, I'm not sure if there is a downside to including the physical properties on all nodes. Probably worth experimenting with this.

> Absolutely. This seems much nicer and easy to understand. Though there may be downsides I'm not seeing right now.

Great. I'm planning to extend this toy a bit more to see if we can capture some of the downsides. Do you have particular transformations that seem complex in the current code? Or a transformation that you'd like to see an example of? One item I'm working on right now is simple decorrelation as that has been a dream for a long time.

> One thing I'm not quite clear how it will work is when we have rendered expressions (e.g. an a+b somewhere in there). Would these show up as special variables in our structures?

Yes, that's exactly what happens. See `opttoy.go:579`. The way this works is that `variableOp` nodes are defined as passing through their input vars to their output vars, but other scalar expressions swallow the output vars. So an expression such as `a+1` in a render expression (i.e. a projection) will have `outputVars == 0` and we'll know that we need to number that expression.

I'm not sure if my handling of `variableOp` is abstracted enough. Currently such expressions only refer to columns, but I think they need to refer to input variables.

Reply:
I don't understand the logical vs physical properties distinction, can you point me to some resource? Are the logical properties those that are true regardless of how we run that node (e.g. equivalency between equality columns in a join) while physical properties depend on the specific way a node is run (e.g. ordering)?

Yes, decorrelation would be a big one. I don't think any of the transformations we are currently doing would be difficult to handle in this model.

Reply:
Yes, that's pretty much the distinction between logical and physical properties as I understand it, though I think I'd restate that as physical properties depend on the specific node. 

One reason for keeping them separate is the Memo structure. The logical properties are the same for all nodes in the same equivalency group while the physical properties are different per node. This can be represented in code by pulling the logical properties out into an `equivalencyGroup` structure and have each `memoExpr` node hold a pointer to the `equivalencyGroup` it is part of. I believe this may also facilitate finding the equivalency group for a node, but I'm still hand-wavy on how that will work. 

There is an overlap between logical and physical properties here that is confusing me right now. Some operators such as `order by` impose physical properties on their output which implies a different equivalency group.

Reply:
Yes, that lat part confuses me as well.

Are there any other things besides ordering that are physical properties? I can't think of anything.

Reply:
Is `unique` a physical property? For a scan operation, seems like we know that based on whether there is a unique index on the set of columns, not whether we actually use that index.

@albler can you weigh in?

Reply:
Even though we may learn about it from a physical aspect of the plan, it's still a logical property in that it's true regardless of how we run it.

Note that "keys" fall under the umbrella of functional dependencies.